### ISSUE TITLE
Proposal: Mount tmpfs on user specified location

### ISSUE THREAD

**Issue by pawelmarkowski** [openfaas/faas#1037]
Hello, 
the only way to mount tmpfs in OpenFaaS is to setup readonlyfs. If you have a lot of IO operations that read/write data then it is not very efficient way to use HDD (or even SSD).

The goal is to move IO operations from hosts HDD to RAMdisk. For example download files directly to tmpfs and perform some calculations on them.

## Expected Behaviour
Add extra option to yaml, where you can specify, which path should be mounted as tmpfs resource for instance --tmpfs /app/mydata/ or even  --tmpfs /tmp/ (without running function on RO mode)  

## Current Behaviour
I use ReadOnlyFS option in OpenFaaS and perform all IO operations in tmp. Unfortunately in swarm and kubernetes there is no option to execute files that are located in /tmp path, which is sometimes necessary. 

existing tickets related to exec /tmp/ 
swarm: https://github.com/moby/moby/issues/32131 
kubernetes: https://github.com/kubernetes/kubernetes/issues/48912

**Comment by bartsmykla** [comment]
Hi @pawelmarkowski. I'm not very familiar with the issue. I'll try to spend some time on it. I don't promise anything yet though. :-)

**Comment by alexellis** [comment]
Hi @bartsmykla please don't work on this until it has design/approve(d) applied.

@pawelmarkowski 

> I use ReadOnlyFS option in OpenFaaS and perform all IO operations in tmp. Unfortunately in swarm and kubernetes there is no option to execute files that are located in /tmp path, which is sometimes necessary.

Can you give steps to reproduce this please?

Why do we want users to execute arbitrary binaries in the writable portion of the disk?

Alex

**Comment by pawelmarkowski** [comment]
@alexellis  --mount tmpfs is a RAMdisk available only in a docker container (let's say that is an alias for --tmpfs from docker run)
as docs [https://docs.docker.com/storage/tmpfs/] says:

> The --tmpfs flag does not allow you to specify any configurable options.
> The --tmpfs flag cannot be used with swarm services. You must use --mount.

In our case: 

1. FaaS downloads big object and associated binaries that are required to properly transform that object (we download it to /tmp that is TMPFS). 
2. Each object has dedicated binaries, which are required to properly transform an object. 
3. We execute that binaries. 
4. After binaries execution we execute our main transform function that performs additional calculations. Without changes made by binaries attached to product our calculations fail.


The case is to allow user to specify - where we want to mount tmpfs (RAM disk). It is required from our perspective because we will destroy extremely fast SSD drives if we do not use TMPFS.

I have to write simple script that will reproduce that steps. Nevertheless it would be great if FaaS can implement such option like --mount type=tmpfs,destination=/app to limit latency caused by IO waits
docker run -d \
  -it \
  --name tmptest \
  --mount type=tmpfs,destination=/app \
  nginx:latest

**Comment by ivanayov** [comment]
@rdimitrov Correct me if I'm wrong, but I think this is in your area of expertise. Could you share your opinion on this?

**Comment by rdimitrov** [comment]
@ivanayov: Unfortunately, it's not, so I'm not sure I can help out on the fly here 😕

**Comment by pawelmarkowski** [comment]
Hello, any news?

**Comment by alexellis** [comment]
So it sounds like this has already been ruled-out for Docker Swarm

 > I believe it meets OpenFaaS principles.

In what way? It doesn't work with Swarm at all.

In Kubernetes @stefanprodan sent this page over which mentions that a tmpfs flag can be used with Kubernetes containers, however this counts against the RAM limitation of the Pod. If you have "huge files" you will need an equal amount of RAM on each host.

https://kubernetes.io/docs/concepts/storage/volumes/

> 1. FaaS downloads big object and associated binaries that are required to properly transform that object (we download it to /tmp that is TMPFS).
> 2. Each object has dedicated binaries, which are required to properly transform an object.
> 3. We execute that binaries.

How often do the big objects and binaries in 1) and 2) change? Have you thought of building them into the container images to save on IO writes?

So to summarise my understanding: Swarm cannot use this proposed feature in any way, even if we wanted to implement it as a feature. Kubernetes can support it, but I'm not convinced we want to enable this as a default, or have enough interest at this point to maintain a feature that only works on Kubernetes for a single end-user.

**Comment by pawelmarkowski** [comment]
> So it sounds like this has already been ruled-out for Docker Swarm

Where can I found information about that? https://docs.docker.com/storage/tmpfs/ (We use tmpfs in docker swarm)

Regarding to your questions:
We use FaaS to transform this objects on user demand. Total size of the object is over 20 PB, so it is not possible to keep them in the docker image. We acquire object on user demand, make proper transformations and save it on dedicated storage. We have rolling archive that removes the oldest objects from that storage.

**Comment by pawelmarkowski** [comment]
Do we have a chance for the feature acceptance or it would be declined?  We have found couple bugs in OpenFaaS, which I will describe in a future tickets but it would be great to know what with our proposal.

**Comment by alexellis** [comment]
Hi @pawelmarkowski if you believe you have found issues with OpenFaaS, then there is a clearly documented process for reporting that. Please follow it and give all the data requested by the template including searching past issues and instructions or an example project to reproduce what you are observing. 

Alex

**Comment by alexellis** [comment]
> In Kubernetes @stefanprodan sent this page over which mentions that a tmpfs flag can be used with Kubernetes containers, however this counts against the RAM limitation of the Pod. If you have "huge files" you will need an equal amount of RAM on each host.

You said that "Total size of the all available objects is over 20 PB"

See the note fro Stefan about the way tmpfs works - it would need a Node with over 20 PB of RAM. I'm unaware of any bare metal or VM instances that have 20 PB of RAM available.

**Comment by pawelmarkowski** [comment]
@alexellis our repository size has over 20PB. We download specified files from that repository and perform some calculations on them. It is continous process with billions IO operations. That is why we are using TMPFS (we download part of data from our repository - between 2-30GB and we make some calculations using OpenFaaS). 

My response regards your question: Have you thought of building them into the container images to save on IO writes? It is not possible because we do not know what part of our repository we need, because we do that jobs "on a customer demand". You are right it is not possible to store 20PB in RAM :)

Regards
